lib/cc/analyzer/include_paths_builder.rb

```diff
@@ -0,0 +1,140 @@
+module CC
+  module Analyzer
+    class IncludePathsBuilder
+      def initialize(cc_excludes)
+        @cc_excludes = cc_excludes
+      end
+
+      def build
+        root = Directory.new('.', included_files)
+        root.included_paths
+      end
+
+      protected
+
+      def ignored_files
+        Tempfile.open(".cc_gitignore") do |tmp|
+          tmp.write(File.read(".gitignore")) if File.exist?(".gitignore")
+          tmp << @cc_excludes.join("\n")
+          tmp.close
+          tracked_and_ignored = `git ls-files -zi -X #{tmp.path}`.split("\0")
+          untracked_and_ignored = `git ls-files -zio -X #{tmp.path}`.split("\0")
+          tracked_and_ignored + untracked_and_ignored
+        end
+      end
+
+      def included_files
+        includable_files - ignored_files
+      end
+
+      def includable_files
+        tracked_in_git = `git ls-files -z`.split("\0")
+        untracked_in_git = `git ls-files -zo`.split("\0")
+        tracked_in_git + untracked_in_git
+      end
+
+      class Directory
+        def initialize(path, included_files)
+          @path = path
+          @included_files = ensure_hashified(included_files)
+        end
+
+        def all_included?
+          readable? && files_all_included? && subdirectories_all_included?
+        end
+
+        def included_paths
+          if all_included?
+            [@path + "/"]
+          elsif readable?
+            result = []
+            result = result.concat(included_file_entries)
+            result = result.concat(included_subdirectory_results)
+            result
+          else
+            []
+          end
+        end
+
+        protected
+
+        def ensure_hashified(obj)
+          if obj.is_a?(Array)
```

Comment:
Basically, there's a necessary step of turning an Array into a Hash as an optimization. In practice it's only run once, at the top directory, and then the hash is passed down recursively.

I opted to do this inside `Directory`, because it felt like the responsibility of `Directory` to know how it's going to use that list and thus optimize it.